server/src/main/java/org/elasticsearch/search/fetch/FetchContext.java

```diff
@@ -0,0 +1,213 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.fetch;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.MapperService;
+import org.elasticsearch.index.query.ParsedQuery;
+import org.elasticsearch.search.SearchExtBuilder;
+import org.elasticsearch.search.fetch.subphase.FetchDocValuesContext;
+import org.elasticsearch.search.fetch.subphase.FetchFieldsContext;
+import org.elasticsearch.search.fetch.subphase.FetchSourceContext;
+import org.elasticsearch.search.fetch.subphase.FieldAndFormat;
+import org.elasticsearch.search.fetch.subphase.InnerHitsContext;
+import org.elasticsearch.search.fetch.subphase.ScriptFieldsContext;
+import org.elasticsearch.search.fetch.subphase.highlight.SearchHighlightContext;
+import org.elasticsearch.search.internal.ContextIndexSearcher;
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.search.rescore.RescoreContext;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Encapsulates state required to execute fetch phases
+ */
+public class FetchContext {
+
+    private FetchContext(SearchContext searchContext) {
+        this.searchContext = searchContext;
+    }
+
+    /**
+     * Create a FetchContext based on a SearchContext
+     */
+    public static FetchContext fromSearchContext(SearchContext context) {
+        return new FetchContext(context);
+    }
+
+    private final SearchContext searchContext;
+
+    /**
+     * The name of the index that documents are being fetched from
+     */
+    public String getIndexName() {
+        return searchContext.indexShard().shardId().getIndexName();
+    }
+
+    /**
+     * The point-in-time searcher the original query was executed against
+     */
+    public ContextIndexSearcher searcher() {
+        return searchContext.searcher();
+    }
+
+    /**
+     * The mapper service for the index we are fetching documents from
+     */
+    public MapperService mapperService() {
+        return searchContext.mapperService();
+    }
+
+    /**
+     * The index settings for the index we are fetching documents from
+     */
+    public IndexSettings getIndexSettings() {
+        return mapperService().getIndexSettings();
+    }
+
+    /**
+     * Gets index field data for a specific fieldtype
+     */
```

Comment:
What do you mean?  A method returning search lookup instead of index field data?

Reply:
Yes!

Reply:
I don't follow 100% how that makes things better but I guess I need to see this updated once #61995 gets merged to understand how the two interact.

Reply:
This is now gone entirely, but I think it would be useful to move SearchLookup directly into the FetchContext - it's not clear at present how to do that and get rid of the new QueryShardContext.newFetchLookup() method though.

Reply:
agreed on moving search lookup into fetch context, I think Nik proposed it as well somewhere.